README.md
# UDP Logger 
![npm](https://img.shields.io/npm/v/udp-logger)
![LGTM Grade](https://img.shields.io/lgtm/grade/javascript/github/JerryCauser/udp-logger)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
![GitHub](https://img.shields.io/github/license/JerryCauser/udp-logger)

Socket and Client for udp logging with possibility to encrypt data.

- Fast — almost zero overhead above UDP to send messages
- Secure — built-in encryption for sensitive logs
- Simple — used well-known Node streams to manipulate and move data 
- Zero-dependency
- ESM and CJS

## Install

```bash
npm i --save udp-logger
```

## Fast Start

```javascript
//app.js
import { logger } from './logger-client.js'

function main () {
  /* ... */
  logger.log('some data to log')
  /* ... */
}

main()
```

```javascript
//logger-client.js
import { UDPLoggerClient } from 'udp-logger'

export const logger = new UDPLoggerClient({ encryption: '11223344556677889900aabbccddeeff' })
```

```javascript
//logger-server.js
import { UDPLoggerServer } from 'udp-logger'

const server = new UDPLoggerServer({
  filePath: './my-app.log',
  decryption: '11223344556677889900aabbccddeeff'
})

await server.start()
```

After just start the logger server `node logger-server.js` and start your app `node app.js`. That's all! Everything is working just fine, and you saved the world 🎉

## Documentation

### class `UDPLoggerClient`
Extends [`EventEmitter`][node-eventemitter]

#### Arguments:
- `options` `<object>` – optional
  - `type` `<'udp4' | 'udp6'>` – optional. **Default** `'udp4'`
  - `port` `<string | number>` – optional. **Default** `44002`
  - `host` `<string>` – optional. **Default** `'127.0.0.1'` or `'::1'`
  - `packetSize` `<number>` – optional. Number of bytes in each packet (chunk). **Default** `1280`
  - `isAsync` `<boolean>` – optional. Enables delayed message sending. Useful if you don't want to wait even for some `ns` in your business flow for logging. **Default** `false`
  - `encryption` `<string> | <(payload: Buffer) => Buffer>` – optional. **Default** `undefined`
    - if passed `string` - will be applied `aes-256-ctr` encryption with passed string as a secret, so it should be 32char long;
    - if passed `function` - will be used that function to encrypt every message;
    - if passed `undefined` - will not use any kind of encryption.
  - `serializer` `<(payload: any) => Buffer>` - optional. **Default** `v8.serialize`
    - Used v8.serialize, but you could use something like [bson][js-bson], or great [avsc][avsc] if you could make all your logs are perfectly typed. Or you even can use `JSON.stringify/JSON.parse` if your logs are simple. It will be faster than v8 serialize/deserialize. Check this [Binary serialization comparison][javascript-serialization-benchmark], it is cool.

#### Methods:
- `log (...args: any)`: `<void>` – just log whatever you need.

#### Events:
- `'ready'`: `<void>` – emitted when the client "establishes" udp connection. But you are not required to wait and just start using `log` method.

#### Usage
##### Simple example with encryption
```javascript
import { UDPLoggerClient } from 'udp-logger'

const logger = new UDPLoggerClient({
  port: 44002,
  // encryption: (buf) => buf.map(byte => byte ^ 83) // not safe at all, but better than nothing
  encryption: '11223344556677889900aabbccddeeff'
})

logger.log('Hello, world!')
logger.log(new Error('Goodbye, world...'))
```

##### Example of several levels of logging with encryption
```javascript
import { UDPLoggerClient } from 'udp-logger'

const encryption = '11223344556677889900aabbccddeeff'

const logs = new UDPLoggerClient({ port: 45001, encryption })
const errors = new UDPLoggerClient({ port: 45002, encryption })
const statistics = new UDPLoggerClient({ port: 45003, encryption })

const logger = {
  log: logs.log,
  error: errors.log,
  stat: statistics.log
}

logger.log('logger created')

export default logger
```

---

### class `UDPLoggerServer`
Extends [`EventEmitter`][node-eventemitter]

It is a simple wrapper around [`UDPLoggerSocket`][socket] and [`UDPLoggerWriter`][writer] created to simplify rapid start

#### Arguments:
- `options` `<object>` – optional
  - `filePath` `<string>` – optional. Supports absolute and relative paths. If passed relative path then will use `process.cwd()` as a base path. **Default** `./udp-port-${port}.log`
  - `encoding` `<string>` — optional. Encoding for writer to your disk. **Default** `'utf8'`,
  - `flags` `<string>` – optional. More info about flags you can check on [NodeJS File System Flags][node-file-system-flags]. **Default** `'a'`
  - `type` `<'udp4' | 'udp6'>` – optional. **Default** `'udp4'`
  - `port` `<string> | <number>` – optional. **Default** `44002`
  - `host` `<string>` – optional **Default** `'127.0.0.1'` or `'::1'`
  - `decryption` `<string> | <(payload: Buffer) => Buffer>` – optional. For more details check [UDPLoggerSocket Arguments Section][socket] **Default** `undefined`
  - `deserializer` `<(payload: Buffer) => any>` - optional. For more details check [UDPLoggerSocket Arguments Section][socket] **Default** `v8.deserialize`
  - `formatMessage` `<(data:any, date:Date, id:number|string) => string | Buffer | Uint8Array>`  - optional. **Default** `DEFAULT_MESSAGE_FORMATTER` from constants

#### Fields: 
- `socket`: `<UDPLoggerSocket>` – instance of [socket][socket].
- `writer`: `<UDPLoggerWriter>` – instance of [writer][writer].

#### Methods:
- `start ()`: `<Promise<UDPLoggerServer>>`
- `stop ()`: `<Promise<UDPLoggerServer>>`

#### Events:
- `'ready'`: `<void>` – emitted when server started
- `'close'`: `<void>` – emitted when server closed
- `'error'`: `<Error~object>` – emitted when some error occurs in `socket` or `writer`. Requires you to handle errors on instance of server.
- `'warning'`: `<any>` – emitted when warning occurs. For more details check [UDPLoggerSocket Events Sections][socket-event-warning]

#### Usage
```javascript
import { UDPLoggerServer } from 'udp-logger'

const server = new UDPLoggerServer({
  filePath: './my-app.log',
  port: 44002,
  // decryption: (buf) => buf.map(byte => byte ^ 83) // not safe at all, but better than nothing
  decryption: '11223344556677889900aabbccddeeff'
})

await server.start()

console.log('log udp server started')
```

---

### class `UDPLoggerSocket`
Extends [`Readable` Stream][node-readable]

It is a UDP socket in `readable stream` form.

#### Arguments:
- `options` `<object>` – **required**
  - `type` `<'udp4' | 'udp6'>` – optional. **Default** `'udp4'`
  - `port` `<string | number>` – optional. **Default** `44002`
  - `host` `<string>` – optional **Default** `'127.0.0.1'` or `'::1'`
  - `decryption` `<string> | <(payload: Buffer) => Buffer>` – optional. **Default** `undefined`
    - if passed `string` - will be applied `aes-256-ctr` decryption with passed string as a secret, so it should be 32char long;
    - if passed `function` - will be used that function to decrypt every message;
    - if passed `undefined` - will not use any kind of decryption.
  - `deserializer` `<(payload: Buffer) => any>` - optional. **Default** `v8.deserialize`
    - Used v8.deserialize, but you could use something like [bson][js-bson], or great [avsc][avsc] if you could make all your logs are perfectly typed. Or you even can use `JSON.stringify/JSON.parse` if your logs are simple. It will be faster than v8 serialize/deserialize. Check this [Binary serialization comparison][javascript-serialization-benchmark], it is cool.
  - `formatMessage` `<(data: any, date:Date, id:number|string) => string | Buffer | Uint8Array>`  - optional. **Default** `DEFAULT_MESSAGE_FORMATTER` from constants

#### Fields:
- `port`: `<number>`
- `address`: `<string>`

#### Events:
All `Readable` events of course and:

##### Event: `'ready'`
Emitted when socket started and ready to receive data.

##### Event: `'message'`
Emitted right after a message was compiled and formatted. Can be used for creating some business logic based on logs (for example pushing some data into google statistics)
  - `message` `<string>` - message
  - `ctx` `<object>`
    - `body` `<any>` - deserialized delivered body
    - `date` `<Date>`
    - `id` `<string>`

##### Event: `'warning'`
Emitted when warning occurs.
 - `payload` `<object | Error>`
   - `message` `<string>`
   - `id` `<string>` – optional
   - `date` `<Date>` – optional

A message might be:
   - `missing_message` – when some messages didn't receive all chunks and got expired.
   - `compile_message_error` – when some messages failed to be compiled from chunks.

#### Usage
##### Example how to use pure socket as async generator
```javascript
import { UDPLoggerSocket } from 'udp-logger'

const socket = new UDPLoggersocket({
  port: 44002,
  // decryption: (buf) => buf.map(byte => byte ^ 83) // not safe at all, but better than nothing
  decryption: '11223344556677889900aabbccddeeff'
})

for await (const message of socket) {
  /*handle messages*/
}
```

##### Example how to use socket and use [UDPLoggerWriter][writer]
```javascript
import { UDPLoggerSocket, UDPLoggerWriter } from 'udp-logger'

const socket = new UDPLoggersocket()
const writer = new UDPLoggerWriter({ filePath: './my-app.log' })

socket.pipe(writer)
```


##### Example with [file-stream-rotator][https://github.com/rogerc/file-stream-rotator]
```javascript
import { UDPLoggerSocket, UDPLoggerWriter } from 'udp-logger'

const socket = new UDPLoggersocket()
const fileStreamWithRotating = (await import('file-stream-rotator'))
        .getStream({ filename: '/tmp/test-%DATE%.log', frequency: 'daily', verbose: false, date_format: 'YYYY-MM-DD' })

socket.pipe(fileStreamWithRotating)
```
---

### class `UDPLoggerWriter`
Extends [`Writable` Stream][node-writable]

`UDPLoggerWriter` handles file moving, renaming, etc.. and keeps working after all. Usable if you will use any kind of logrotating ([linux logrotate][linux-logrotate] for example) 

#### Arguments:
- `options` `<object>` – **required**
  - `filePath` `<string>` – **required**. Supports absolute and relative paths. If passed relative path then will use `process.cwd()` as a base path
  - `encoding` `<string>` — optional. Encoding for writer to your disk. **Default** `'utf8'`
  - `flags` `string` – optional. More info about flags you can check on [NodeJS File System Flags][node-file-system-flags]. **Default** `'a'`

#### Events:
- `'ready'`: `<void>` – emitted when write stream ready

 
#### Usage
##### Example how to use socket and use [UDPLoggerSocket][socket]

```javascript
import { UDPLoggerSocket, UDPLoggerWriter } from 'udp-logger'

const socket = new UDPLoggersocket()
const writer = new UDPLoggerWriter({ filePath: './my-app.log' })

socket.pipe(writer)
```
---


### Additional Exposed variables and functions from [`constants`][constants]
#### function `DEFAULT_MESSAGE_FORMATTER(data, date, id)`
 - `data` `<any[]>` — deserialized arguments of `log` function from [`UDPLoggerClient`][client]
 - `date` `<Date>`
 - `id` `<string>`
 - Returns: `<Buffer>` | `<string>` | `<Uint8Array>`

Default format function to write logs. Underhood it uses [util-inspect][node-util-inspect] function. For more info, you can check the source files.

#### constant `DEFAULT_PORT`
- `<number>` : `44002`
---

License ([MIT](LICENSE))

[js-bson]: https://github.com/mongodb/js-bson
[avsc]: https://github.com/mtth/avsc
[javascript-serialization-benchmark]: https://github.com/Adelost/javascript-serialization-benchmark
[node-file-system-flags]: https://nodejs.org/api/fs.html#file-system-flags
[node-eventemitter]: https://nodejs.org/api/events.html#class-eventemitter
[node-readable]: https://nodejs.org/api/stream.html#class-streamreadable
[node-writable]: https://nodejs.org/api/stream.html#class-streamwritable
[node-util-inspect]: https://nodejs.org/api/util.html#utilinspectobject-options
[client]: #class-udploggerclient
[server]: #class-udploggerserver
[socket]: #class-udploggersocket
[writer]: #class-udploggerwriter
[constants]: src/constants.js
[socket-event-warning]: #event-warning
[linux-logrotate]: https://linux.die.net/man/8/logrotate


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.5.0](https://github.com/JerryCauser/udp-logger/compare/v0.4.0...v0.5.0) (2022-07-14)


### Features

* add sync/async behavior on client ([b1b8d0f](https://github.com/JerryCauser/udp-logger/commit/b1b8d0f23dadef44fb65ae071e36ef3c34f1baa6))
* detach encoding function and add possibility to pass encode/decode functions instead of algo names ([a6087e2](https://github.com/JerryCauser/udp-logger/commit/a6087e2cfc44739a5e5d1afbbe185e78b3ee55c1))

## [0.4.0](https://github.com/JerryCauser/udp-logger/compare/v0.3.0...v0.4.0) (2022-05-18)


### Features

* improve buffering messages in socket ([6d5f6d2](https://github.com/JerryCauser/udp-logger/commit/6d5f6d25c415f924514f0bbed11111c2d255dd38))


### Bug Fixes

* server error on end ([21beb4c](https://github.com/JerryCauser/udp-logger/commit/21beb4ca666ff6f47c850d8726d1dc71e2d93c35))

## [0.3.0](https://github.com/JerryCauser/udp-logger/compare/v0.2.0...v0.3.0) (2022-05-17)


### ⚠ BREAKING CHANGES

* node version >= v15.0.0 allowed

Because of using new `node:stream` `_construct` API the minimum node engine version now 15.0.0

### Features

* split server on socket and server ([cae5e62](https://github.com/JerryCauser/udp-logger/commit/cae5e62e294e6196f39431a8b5107a3ec1658814))


### Bug Fixes

* style lint ([38eaf73](https://github.com/JerryCauser/udp-logger/commit/38eaf7305f0c870e189bee0145c78ac780b10cd4))


* split server to socket and writer ([3920e0e](https://github.com/JerryCauser/udp-logger/commit/3920e0e06352aac182c3b6369d6b9a19bb2c5114))

## 0.2.0 (2022-04-26)


src/socket.js
import EventEmitter from 'node:events'
import dgram from 'node:dgram'
import { Buffer } from 'node:buffer'
import { Readable } from 'node:stream'
import { ID_SIZE, parseId } from './identifier.js'
import {
  DEFAULT_PORT,
  DEFAULT_MESSAGE_FORMATTER,
  DEFAULT_DESERIALIZER,
  DEFAULT_DECRYPT_FUNCTION
} from './constants.js'

/**
 * @typedef {object} UDPLoggerSocketOptions
 * @property {string} [type='udp4']
 * @property {number} [port=44002]
 * @property {string} [host=('127.0.0.1'|'::1')]
 * @property {string | ((payload: Buffer) => Buffer)} [decryption]
 *    if passed string - will be applied aes-256-ctr encryption with passed string as secret, so it should be 32char long;
 *    if passed function - will be used that function to encrypt every message;
 *    if passed nothing - will not use any kind of encryption
 * @property {(payload: Buffer) => any} [deserializer]
 * @property {(data: any, date:Date, id:number|string) => string | Buffer | Uint8Array} [formatMessage]
 *
 * @extends {ReadableOptions}
 */

/**
 * @class
 * @param {UDPLoggerSocketOptions} [options={}]
 */
class UDPLoggerSocket extends Readable {
  /** @type {string} */
  #host

  /** @type {number} */
  #port

  /** @type {string} */
  #address

  /** @type {'udp4'|'udp6'} */
  #type

  /** @type {undefined|function(Buffer): Buffer} */
  #decryptionFunction

  /** @type {undefined|string} */
  #decryptionSecret

  /** @type {dgram.Socket} */
  #socket

  /** @type {(Buffer) => any} */
  #deserializer

  /** @type {(data: any, date:Date, id:number|string) => string | Buffer | Uint8Array} */
  #formatMessage

  /** @type {Map<string, [logBodyMap:Map, lastUpdate:number, logDate:Date, logId:string]>} data */
  #collector = new Map()

  /** @type {number} */
  #gcIntervalId

  /** @type {number} */
  #gcIntervalTime = 5000

  /** @type {number} */
  #gcExpirationTime = 10000

  /** @type {boolean} */
  #allowPush = true

  /** @type {(string | Buffer | Uint8Array)[]} */
  #messages = []

  /** @type {function (Buffer):void} */
  #handleSocketMessage

  /**
   * @param {UDPLoggerSocketOptions} [options]
   */
  constructor ({
    type = 'udp4',
    port = DEFAULT_PORT,
    host = type === 'udp4' ? '127.0.0.1' : '::1',
    decryption,
    deserializer = DEFAULT_DESERIALIZER,
    formatMessage = DEFAULT_MESSAGE_FORMATTER,
    ...readableOptions
  } = {}) {
    super({ ...readableOptions })

    this.#port = port
    this.#deserializer = deserializer
    this.#formatMessage = formatMessage
    this.#type = type

    if (decryption) {
      if (typeof decryption === 'string') {
        this.#decryptionSecret = Buffer.from(decryption, 'hex')

        this.#decryptionFunction = (data) =>
          DEFAULT_DECRYPT_FUNCTION(data, this.#decryptionSecret)
      } else if (decryption instanceof Function) {
        this.#decryptionFunction = decryption
      }
    }

    this.#handleSocketMessage = this.#handlePlainMessage

    if (this.#decryptionFunction instanceof Function) {
      this.#handleSocketMessage = this.#handleEncryptedMessage
    }
  }

  _construct (callback) {
    this.#start()
      .then(() => callback(null))
      .catch(callback)
  }

  _destroy (error, callback) {
    if (error) {
      this.emit('error', error)
    }

    this.#stop()
      .then(() => callback(error))
      .catch(callback)
  }

  _read (size) {
    this.#sendBufferedMessages()

    this.#allowPush = this.#messages.length === 0
  }

  /**
   * @param {*} message
   */
  #addMessage (message) {
    if (this.#allowPush) {
      this.#allowPush = this.push(message)
    } else {
      this.#messages.push(message)
    }
  }

  #sendBufferedMessages () {
    if (this.#messages.length === 0) return

    for (let i = 0; i < this.#messages.length; ++i) {
      if (!this.push(this.#messages[i])) {
        this.#messages.splice(0, i + 1)
        break
      }
    }
  }

  get address () {
    return this.#address
  }

  get port () {
    return this.#port
  }

  async #start () {
    this.#collector.clear()
    this.#gcIntervalId = setInterval(this.#gcFunction, this.#gcIntervalTime)
    await this.#initSocket()
    this.#attachHandlers()

    this.#address = this.#socket.address().address
    this.#port = this.#socket.address().port

    this.emit('ready')
  }

  async #stop () {
    clearInterval(this.#gcIntervalId)
    this.#collector.clear()

    if (!this.#socket) {
      return
    }

    this.#detachHandlers()
    this.#socket.close()

    await EventEmitter.once(this.#socket, 'close')
  }

  async #initSocket () {
    this.#socket = dgram.createSocket({ type: this.#type })
    this.#socket.bind(this.#port, this.#host)

    const error = await Promise.race([
      EventEmitter.once(this.#socket, 'listening'),
      EventEmitter.once(this.#socket, 'error')
    ])

    if (error instanceof Error) {
      this.destroy(error)
    }
  }

  #attachHandlers () {
    this.#socket.on('error', this.#handleSocketError)
    this.#socket.on('message', this.#handleSocketMessage)
  }

  #detachHandlers () {
    this.#socket.off('error', this.#handleSocketError)
    this.#socket.off('message', this.#handleSocketMessage)
  }

  #handleSocketError = (error) => {
    this.destroy(error)
  }

  /**
   * @param {Buffer} buffer
   */
  #handlePlainMessage = (buffer) => {
    const [date, id, total, index] = parseId(buffer.subarray(0, ID_SIZE))

    /** @type {[logBodyMap:Map, lastUpdate:number, logDate:Date, logId:string]} */
    let data = this.#collector.get(id)
    if (!data) {
      data = [new Map(), Date.now(), date, id]
      this.#collector.set(id, data)
    }

    data[0].set(index, buffer.subarray(ID_SIZE))

    if (data[0].size === total + 1) {
      this.#collector.delete(id)
      this.#compileMessage(data[0], data[2], data[3])
    }
  }

  /**
   * @param {Buffer} buffer
   */
  #handleEncryptedMessage = (buffer) => {
    try {
      return this.#handlePlainMessage(this.#decryptionFunction(buffer))
    } catch (e) {
      console.error(e)
    }
  }

  #gcFunction = () => {
    const dateNow = Date.now()

    for (const [id, payload] of this.#collector) {
      if (payload[1] + this.#gcExpirationTime < dateNow) {
        this.#collector.delete(id)
        this.emit('warning', {
          message: 'missing_message',
          id,
          date: payload[2]
        })
      }
    }
  }

  /**
   * @param {Map<number, Buffer>} body
   * @param {Date|number} date
   * @param {string} id
   */
  #compileMessage (body, date, id) {
    try {
      this.#compileMessageUnsafe(body, date, id)
    } catch (error) {
      const originMessage = error.message

      error.message = 'compile_message_error'
      error.ctx = {
        originMessage,
        date,
        id
      }

      this.emit('warning', error)
    }
  }

  /**
   * @param {Map<number, Buffer>} body
   * @param {Date|number} date
   * @param {string} id
   */
  #compileMessageUnsafe (body, date, id) {
    let bodyBuffered

    if (body.size > 1) {
      const sortedBuffers = [...body.entries()]
        .sort((a, b) => a[0] - b[0])
        .map((n) => n[1])

      bodyBuffered = Buffer.concat(sortedBuffers)
    } else {
      bodyBuffered = [...body.values()][0]
    }

    const deserializedBody = this.#deserializer(bodyBuffered)
    const message = this.#formatMessage(deserializedBody, date, id)

    this.#addMessage(message)

    this.emit('message', message, { body: deserializedBody, date, id })
  }
}

export default UDPLoggerSocket
